libs/cardano-data/src/Data/Incremental.hs

```diff
@@ -0,0 +1,181 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- | Introduce the Incremental Lambda Calculus embodied in the ILC class.
+--   Instances for two patterns of use involving Maps.
+module Data.Incremental where
+
+import Control.DeepSeq (NFData (..))
+import Data.Kind
+import Data.Map.Internal (Map (..))
+import Data.Map.Strict
+import qualified Data.Map.Strict as Map
+import GHC.Generics (Generic (..))
+
+-- ===================================================
+-- Incremental lambda calculus
+
+class ILC t where
+  data Diff t :: Type
+  applyDiff :: t -> Diff t -> t
+  extend :: Diff t -> Diff t -> Diff t
+  zero :: Diff t
+
+infixr 0 $$
+($$) :: ILC t => t -> Diff t -> t
+x $$ y = applyDiff x y
+
+-- | Every (Diff t) is a Semigroup
+instance ILC t => Semigroup (Diff t) where
+  x <> y = extend x y
+
+-- | Every (Diff t) is a Monoid
+instance ILC t => Monoid (Diff t) where
+  mempty = zero
+
+-- ========================================
+-- Delta types
+
+-- | Its is deleted, overwritten, or combined using a Monoid
+data MonoidD v = Del | Write !v | Comb !v
+  deriving (Show, Eq, Generic, NFData)
+
+instance (Semigroup t) => Semigroup (MonoidD t) where
+  Del <> Del = Del
+  Del <> Write _ = Del
+  Del <> Comb _ = Del
+  Comb x <> Del = Write x
+  Comb x <> Write y = Write (x <> y)
+  Comb x <> Comb y = Comb (x <> y)
+  Write x <> Del = Write x
+  Write x <> Comb _ = Write x
+  Write x <> Write _ = Write x
+
+-- | It is deleted or changed
+data BinaryD v = Omit | Edit !v
+  deriving (Eq, Generic, NFData)
+
+instance Show v => Show (BinaryD v) where
+  show Omit = "Omit"
+  show (Edit d) = "Edit(" ++ show d ++ ")"
+
+instance Semigroup (BinaryD t) where
+  Omit <> Omit = Omit
+  Omit <> Edit _ = Omit
+  Edit x <> Omit = Edit x
+  Edit x <> Edit _ = Edit x
+
+-- ======================================================
+-- Special case of a Map where the range is a Monoid
+-- we provide tools to enforce the invariant, that in a
+-- MonoidMap, we never store  'mempty' of the Monoid
+
+newtype MonoidMap k v = MM (Map k v)
+  deriving newtype (Eq, NFData)
+
+unMM :: MonoidMap k v -> Map k v
+unMM (MM x) = x
+
+monoidInsertWith :: (Monoid v, Eq v, Ord k) => k -> v -> MonoidMap k v -> MonoidMap k v
+monoidInsertWith k !v1 (MM m) = MM (alter ok k m)
+  where
+    ok Nothing = if v1 == mempty then Nothing else Just v1
+    ok (Just v2) = if total == mempty then Nothing else Just total
+      where
+        total = v1 <> v2
+{-# INLINEABLE monoidInsertWith #-}
+
+monoidInsert :: (Monoid v, Eq v, Ord k) => k -> v -> MonoidMap k v -> MonoidMap k v
+monoidInsert k v1 (MM m) = if v1 == mempty then MM (delete k m) else MM (insert k v1 m)
```

Comment:
How come in `monoidInsertWith` above we want v to be strict, but not here? On purpose?

Reply:
No it is an oversight. Good catch.